R/grade_this_learnr.R

```diff
@@ -0,0 +1,286 @@
+        do.call(parse_checker, append(list(parse_error = e), learnr_args))
+      })
+    }
+  )
+
+  if (is_graded(user_code)) {
+    user_code <- feedback(user_code)
+  }
+  if (is_feedback(user_code)) {
+    return(user_code)
+  }
+  if (length(user_code) == 0) {
+    return(feedback(
+      fail("I didn't receive your code. Did you write any?"),
+      type = "info"
+    ))
+  }
+
+  # Sometimes no solution is provided, but that
+  # means there is nothing to check against. Also,
+  # you do not want to parse NULL
+  if (!is.null(solution_code)) {
+    solution_code <- parse(text = solution_code)
+    if (length(solution_code) == 0) {
+      return(feedback(
+        fail("No solution is provided for this exercise."),
+        type = "info"
+      ))
+    }
+  }
+
+  check_label <-
+    if (is.null(envir_result)) {
+      paste0(label, "-code-check")
+    } else {
+      paste0(label, "-check")
+    }
+
+  ## setup an environment for checking
+  # envir for function call
+  chunk_envir <- learnr::duplicate_env(envir_prep)
+  # object to pass to checking function that should contain all of the information
+  check_obj_envir <- new.env(parent = chunk_envir)
+
+  # Copy over all learnr args into the checking environment
+  for (name in names(learnr_args)) {
+    check_obj_envir[[paste0(".", name)]] <- learnr_args[[name]]
+  }
+
+  # Add gradethis specific check objects
+  check_obj_envir[[".result"]] <- last_value
+  # check_obj_envir[["."]] <- last_value # keep this one? No
+  # check_obj_envir[[".user"]] <- last_value
+  # check_obj_envir[[".user"]] <- rlang::as_quosure(user_code, envir_result)
+  delayedAssign(
+    assign.env = check_obj_envir,
+    x = ".solution",
+    {
+      message("Delayed evaluation of `.solution!`")
+      # (empty solutioncode is checked above)
+      # solution code exists...
+      rlang::eval_tidy(
+        rlang::as_quosure(solution_code, envir_result)
+      )
+    }
+  )
+
+  # evaluate all checking code
+  checking_envir <- rlang::current_env()
+  to_check_fn <- capture_errors(
+    {
+      parsed_check_code <- parse(text = check_code %||% "")
+      capture_graded(
+        {
+          eval(parsed_check_code, envir = chunk_envir)
+        },
+        # if a `pass()`/`fail()` is used in the regular check chunk with no user submission context, it should be an error
+        on_graded = function(gc_obj, ignore) {
+          fn_used <- if (isTRUE(gc_obj$correct)) "`pass()`" else "`fail()`"
+
+          # notify author of their mistake
+          message(
+            "A ", fn_used, " statement was executed without access to student feedback. (Prematurely graded)\n",
+            "Remember to only call ", fn_used, " inside your checking function (ex: `grade_this({})`"
+          )
+          # return from main function (even though in a inner function! voodoo!)
+          rlang::return_from(checking_envir, feedback(
+            fail("Uh Oh! Submission prematurely graded. Marking as _incorrect_"),
```

Comment:
```suggestion
            fail("A problem occurred with your teacher's grading code. Defaulting to _incorrect_"),
```

Reply:
Copy this in for all similar instances where it is the grading code that causes the error.